### PR TITLE
Constrain services to only run on controllers when doing HA

### PR DESCRIPTION
**Pull request by vuntz**
This depends on ~~https://github.com/crowbar/crowbar-openstack/pull/175~~ and ~~https://github.com/crowbar/crowbar-ha/pull/56~~

### CODE REVIEW

**Comment by jsuchome** [comment]
I haven't tested it, but the code makes sense to me. +1

**Comment by aplanas** [comment]
+1 (gating fail not related)

**Comment by vuntz** [comment]
One tempest test is causing the gating to fail, and it's happening because of:

```
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [req-010eb07d-b911-4cfa-96ae-fa71197cc014 c5b7e829399c48b6b63a9c2ec4b1c24d 6745fdf7fd14483ab0bce0efc0004650 - - -] [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb] Instance failed to spawn
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb] Traceback (most recent call last):
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]   File "/usr/lib/python2.7/site-packages/nova/compute/manager.py", line 2155, in _build_resources
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]     yield resources
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]   File "/usr/lib/python2.7/site-packages/nova/compute/manager.py", line 2009, in _build_and_run_instance
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]     block_device_info=block_device_info)
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]   File "/usr/lib/python2.7/site-packages/nova/virt/libvirt/driver.py", line 2437, in spawn
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]     admin_pass=admin_password)
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]   File "/usr/lib/python2.7/site-packages/nova/virt/libvirt/driver.py", line 2936, in _create_image
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]     with configdrive.ConfigDriveBuilder(instance_md=inst_md) as cdb:
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]   File "/usr/lib/python2.7/site-packages/nova/virt/configdrive.py", line 72, in __init__
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]     self.add_instance_metadata(instance_md)
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]   File "/usr/lib/python2.7/site-packages/nova/virt/configdrive.py", line 93, in add_instance_metadata
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]     for (path, data) in instance_md.metadata_for_config_drive():
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]   File "/usr/lib/python2.7/site-packages/nova/api/metadata/base.py", line 465, in metadata_for_config_drive
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]     yield (filepath, jsonutils.dump_as_bytes(data['meta-data']))
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb] AttributeError: 'module' object has no attribute 'dump_as_bytes'
2016-01-08 09:48:02.961 14849 ERROR nova.compute.manager [instance: 2607a025-00ee-4f94-8616-a5ca3d2cd2cb]
```

**Comment by vuntz** [comment]
> One tempest test is causing the gating to fail, and it's happening because of:

It's a bug in nova, now fixed in the package.

**Comment by aspiers** [comment]
Hmmm ... I really dislike the amount of duplication introduced by this.  I thought we agreed we were going to use a Chef definition to avoid it?

**Comment by vuntz** [comment]
> Hmmm ... I really dislike the amount of duplication introduced by this. I thought we agreed we were going to use a Chef definition to avoid it?

I thought about that again, but I'm not convinced it's worth it anymore. We would be  2 (maybe 3) lines shorter, with the downside of making the thing a bit less readable (imho) when you look at the creation of a full transaction: the nice thing right now is that you can read each block about a pacemaker resource and not be surprised by the fact that it's going to be a resource created in the transaction.

(also, definitions are being deprecated in chef)

Just so we're all on the same page and can all share an opinion, with a definition, it would look like:

```ruby
location_name = "l-#{clone_name}-controller"
openstack_pacemaker_controller_only_location location_name
  resource clone_name
end
transaction_objects << "pacemaker_location[#{location_name}]"
```

instead of

```ruby
location_name = "l-#{clone_name}-controller"
pacemaker_location location_name do
  definition OpenStackHAHelper.controller_only_location(location_name, clone_name)
  action :update
  only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
end
transaction_objects << "pacemaker_location[#{location_name}]"
```

Hrm, thinking about it, we could possibly make it look like:
```ruby
location_name = openstack_pacemaker_controller_only_location_for clone_name
transaction_objects << "pacemaker_location[#{location_name}]"
```

Would be something like:
```ruby
define :openstack_pacemaker_controller_only_location_for
  resource = params[:name]
  location_name = "l-#{resource}-controller"
  pacemaker_location location_name do
    definition OpenStackHAHelper.controller_only_location(location_name, resource)
    action :update
    only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
  end
end
```

**Comment by vuntz** [comment]
@aspiers see last two commits.

**Comment by vuntz** [comment]
Manual testing shows that this works as intended.

File: chef/cookbooks/rabbitmq/recipes/ha.rb [hunk 1]
```diff
@@ -294,6 +300,17 @@
...
+  admin_vip_location_name = openstack_pacemaker_controller_only_location_for admin_vip_primitive
+  service_transaction_objects << "pacemaker_location[#{admin_vip_location_name}]"
+
+  if node[:rabbitmq][:listen_public]
+    public_vip_location_name = openstack_pacemaker_controller_only_location_for public_vip_primitive
+    service_transaction_objects << "pacemaker_location[#{public_vip_location_name}]"
+  end
+
+  location_name = openstack_pacemaker_controller_only_location_for service_name
+  service_transaction_objects << "pacemaker_location[#{location_name}]"
```
Comment (vuntz) [inline chef/cookbooks/rabbitmq/recipes/ha.rb]: I beg to differ, we try to keep these lines grouped.

**Comment by jsuchome** [comment]
+1 after looking at 4 latest commits

**Comment by aplanas** [comment]
Re-review: +1

**Comment by nkrinner** [comment]
+1

**Comment by vuntz** [comment]
FWIW, I force-pushed after @aplanas and @jsuchome latest +1, but only to change commit messages (based on feedback from @aspiers) and to add the last commit.

@aspiers do you want to get more clarifications in the messages, or are we good to merge?

**Comment by aspiers** [comment]
+1.  Again I would have preferred some commits to be squashed (especially since you add a lot of new code and only then make a definition to remove that duplication again; would have been less to review if the helper had been fixed first).  But let's just move on in the interests of time.

**Pull request merged**